reframe/core/pipeline.py

```diff
@@ -760,6 +760,22 @@ def __new__(cls, *args, _rfm_use_params=False, **kwargs):
     def __init__(self):
         pass
 
+    def __getattribute__(self, name):
+        try:
+            return super().__getattribute__(name)
+
+        except AttributeError as err:
+            # Intercept the AttributeError if the name corresponds to a
+            # required variable.
+            if (name in self._rfm_var_space.vars and
+                not self._rfm_var_space.vars[name].is_defined()):
+                raise AttributeError(
+                    f'required variable {name!r} has not been defined'
+                ) from None
+
+            else:
+                super().__getattr__(name)
```

Comment:
You are right, because this way you respect the `__getattr__` semantics. My comment was because you were already handling the `AttributeError` from `__getattribute__`, so I was like how could `__getattr__` raise it again. But you are right, the superclass might define it, so we should call it.